test/extended/devices/recovery.go

```diff
@@ -0,0 +1,420 @@
+package devices
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"strings"
+	"time"
+
+	g "github.com/onsi/ginkgo/v2"
+	o "github.com/onsi/gomega"
+
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/resource"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+	kubeletdevicepluginv1beta1 "k8s.io/kubelet/pkg/apis/deviceplugin/v1beta1"
+	"k8s.io/kubernetes/test/e2e/framework"
+	e2eskipper "k8s.io/kubernetes/test/e2e/framework/skipper"
+	admissionapi "k8s.io/pod-security-admission/api"
+
+	exutil "github.com/openshift/origin/test/extended/util"
+	"github.com/openshift/origin/test/extended/util/image"
+)
+
+const (
+	workloadCommand = "devs=$(ls /tmp/ | egrep '^Dev-[0-9]+$') || { echo MISSING; exit 1; } && { echo devices=$devs; sleep 24h; }" // sleep forever in the test timescale
+
+	setupRegistrationCommandMCD = "mkdir -p /rootfs/var/lib/kubelet/device-plugins/sample && touch /rootfs/var/lib/kubelet/device-plugins/sample/registration"
+	// it's just a file plus just a directory. We reverse the step to be on the safer side and minimize the negative effects on bugs.
+	// note the file is expected to be gone, so we must tolerate the rm to fail. In any case, rmdir should never fail.
+	cleanupRegistrationCommandMCD = "rm /rootfs/var/lib/kubelet/device-plugins/sample/registration; rmdir /rootfs/var/lib/kubelet/device-plugins/sample"
+	unblockRegistrationCommandMCD = "rm /rootfs/var/lib/kubelet/device-plugins/sample/registration"
+	rebootNodeCommandMCD          = "chroot /rootfs systemctl reboot"
+
+	sampleDeviceResourceName            = "example.com/resource"
+	sampleDevicePluginPodName           = "sample-device-plugin"
+	sampleDeviceEnvVarNamePluginSockDir = "PLUGIN_SOCK_DIR"
+
+	sampleDeviceEnvVarNameControlFile = "REGISTER_CONTROL_FILE"
+	sampleDevicePluginControlFilePath = "/var/lib/kubelet/device-plugins/sample/registration"
+
+	sampleDevicePluginPodDSYAML = "test/e2e/testing-manifests/sample-device-plugin/sample-device-plugin-control-registration.yaml"
+
+	targetNodeEnvVar = "DEVICE_RECOVERY_TARGET_NODE"
```

Comment:
where does the actual setup of the DEVICE_RECOVERY_TARGET_NODE env var happens ? in selective tests ?  
will the test suits running for this PR actually include that or just skip the tests?

Reply:
The entity (human or program) running the test is in charge of setting the environment variable, otherwise the test will always skip. Thing is, didn't find obvious way to pick a good target automatically, so having the explicit selection requirement seem the lesser evil. I'm open to suggestions though.